ta/pkcs11/src/processing_aes.c

```diff
@@ -14,6 +15,385 @@
+			goto bail;
+		}
+		ctx->out_data = ptr;
+		ctx->out_data[ctx->out_count].size = ct_size;
+		ctx->out_data[ctx->out_count].data = ct;
+		ctx->out_count++;
+	}
+
+	rc = PKCS11_CKR_OK;
+
+bail:
+	if (rc)
+		TEE_Free(ct);
+
+	return rc;
+}
+
+static enum pkcs11_rc reveale_ae_data(struct ae_aes_context *ctx,
+				      void *out, size_t *out_size)
+{
+	size_t n = 0;
+	uint32_t req_size = 0;
+	char *out_ptr = out;
+
+	for (req_size = 0, n = 0; n < ctx->out_count; n++)
+		req_size += ctx->out_data[n].size;
+
+	if (*out_size < req_size) {
+		*out_size = req_size;
+		return PKCS11_CKR_BUFFER_TOO_SMALL;
+	}
+
+	if (!out_ptr)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	for (n = 0; n < ctx->out_count; n++) {
+		TEE_MemMove(out_ptr,
+			    ctx->out_data[n].data, ctx->out_data[n].size);
+
+		TEE_Free(ctx->out_data[n].data);
+		out_ptr += ctx->out_data[n].size;
+	}
+
+	TEE_Free(ctx->out_data);
+	ctx->out_data = NULL;
+	ctx->out_count = 0;
+
+	*out_size = req_size;
+
+	return PKCS11_CKR_OK;
+}
+
+enum pkcs11_rc tee_ae_decrypt_final(struct pkcs11_session *session,
+				    void *out, size_t *out_size)
+{
+	struct ae_aes_context *ctx = session->processing->extra_ctx;
+	enum pkcs11_rc rc = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t data_size = 0;
+	void *data_ptr = NULL;
+
+	if (!out_size) {
+		DMSG("Expect at least a buffer for the output data");
+		return PKCS11_CKR_ARGUMENTS_BAD;
+	}
+
+	/* Final is already completed, only need to output the data */
+	if (!ctx->pending_tag)
+		return reveale_ae_data(ctx, out, out_size);
+
+	if (ctx->pending_size != ctx->tag_byte_len) {
+		DMSG("Not enough samples: %zu/%zu",
+		     ctx->pending_size, ctx->tag_byte_len);
+		return PKCS11_CKR_ENCRYPTED_DATA_LEN_RANGE;
+	}
+
+	data_size = 0;
+	res = TEE_AEDecryptFinal(session->processing->tee_op_handle,
+				 NULL, 0, NULL, &data_size,
+				 ctx->pending_tag, ctx->tag_byte_len);
+
+	if (res == TEE_ERROR_SHORT_BUFFER) {
+		data_ptr = TEE_Malloc(data_size,
+				      TEE_USER_MEM_HINT_NO_FILL_ZERO);
+		if (!data_ptr) {
+			rc = PKCS11_CKR_DEVICE_MEMORY;
+			goto bail;
+		}
+
+		res = TEE_AEDecryptFinal(session->processing->tee_op_handle,
+					 NULL, 0, data_ptr, &data_size,
+					 ctx->pending_tag, ctx->tag_byte_len);
+
+		if (!data_size) {
+			TEE_Free(data_ptr);
+			data_ptr = NULL;
+			DMSG_RAW("Is this expected from the Core API?\n\n");
+		}
+	}
+
+	/* AE decryption is completed */
+	TEE_Free(ctx->pending_tag);
+	ctx->pending_tag = NULL;
+
+	rc = tee2pkcs_error(res);
+	if (rc)
+		goto bail;
+
+	if (data_ptr) {
+		void *tmp_ptr = NULL;
+
+		tmp_ptr = TEE_Realloc(ctx->out_data,
+				      (ctx->out_count + 1) *
+					sizeof(struct out_data_ref));
+		if (!tmp_ptr) {
+			rc = PKCS11_CKR_DEVICE_MEMORY;
+			goto bail;
+		}
+		ctx->out_data = tmp_ptr;
+		ctx->out_data[ctx->out_count].size = data_size;
+		ctx->out_data[ctx->out_count].data = data_ptr;
+		ctx->out_count++;
+
+		data_ptr = NULL;
+	}
+
+	rc = reveale_ae_data(ctx, out, out_size);
+
+bail:
```

Comment:
please replace all `bail` branch labels with `out` branch labels.
I remember @jenswi-linaro asked me this at the time when we reviewed the pkcs11 TA implementation.